### ISSUE TITLE
Cobertura 2.0.3 - upgrading Groovy from 2.0.4 to 2.3.6 drops coverage from 98% to 49%

### ISSUE THREAD

**Issue by mecorusfc** [cobertura/cobertura#184]
(This is Java 7 by the way.) Many situations cause coverage to not be recorded once really anything beyond Groovy 2.0.4 is used.  All the 2.1.x's I've tested get very low coverage.  Also, 2.2's.  However, I had high hope when I started moving utility jar projects to 2.3.6.  It seemed that if I made sure not to have a pattern where I didn't specify the return on a signature, it could get the coverage.  Here is an example:

protected someMethodName() { do stuff here } (no coverage recorded no matter what I do)

protected void someMethodName() { do stuff here } - coverage generally recorded ok

Another issue is this:

if (a == b) {
   do stuff here
} else {
   do different stuf
}

It will show that the if is hit 40 times.  It will show that the first branch is hit 5 times.  It will then show that the second else is not covered at all with a zero.  This is very common when moving from 2.0.4 to anything higher in Groovy.

Note that sometimes the disableoptimizationstransformation jar helps but in my current situation, it makes no difference at all.

Here is another example of the same exact code and tests between Groovy 2.0.4 and any higher version, although 2.3.6 seems to get closer:
## Groovy 2.0.4

@Override
 40              void initialize(Bootstrap<ApplicationConfiguration> bootstrap) {
 41  5                   bootstrap.setName('some-service')
 42  5                   bootstrap.objectMapperFactory.registerModule(new JaxbAnnotationModule())
 43              }
## Groovy 2.3.6

```
     @Override
```

 40              void initialize(Bootstrap<ApplicationConfiguration> bootstrap) {
 41  0                   bootstrap.setName('some-service')
 42  0                   bootstrap.objectMapperFactory.registerModule(new JaxbAnnotationModule())
 43              }

This is a lot of versions of Groovy for Cobertura not to be keeping up.  Is there any testing going on with Groovy at all?  It seems there may not be.  I can provide many more examples. 

I really need to get a working version of Cobertura.  Is there a nightly build version that solves these issues?

Mike

**Comment by samrocketman** [comment]
I can see this illustrated in my project building on Travis CI.  See this Travis build in which it tests Groovy version 1.8.9 through 2.4.0.

https://travis-ci.org/samrocketman/jervis/builds/49014379

In my `.travis.yml`:

    script: ./gradlew coberturaCheck test

Groovy 1.8.9 and Groovy 2.0.8 pass code coverage tests.  However, Groovy 2.1.9 and later fails.  To generate code coverage reports with different versions of Groovy do the following.

    GROOVY_VERSION="1.8.9" ./gradlew cobertura
    GROOVY_VERSION="2.4.0" ./gradlew cobertura

**Comment by dennisl** [comment]
Could you please try using the Cobertura 2.1.1 release?

**Comment by blackdrag** [comment]
where can you download Cobertura 2.1.1? The web page states 2.0.3 is the latest version (which I think is from 2013)

**Comment by dennisl** [comment]
I've added a download section to the README at
https://github.com/cobertura/cobertura

**Comment by blackdrag** [comment]
I suggest you update the website as well a putting a release on sourceforge

**Comment by samrocketman** [comment]
@dennisl Using the latest cobertura 2.1.1 release causes 100% of my groovy tests to fail.  It seems to be doing something bad to the runtime.  I get errors like:

```
jervis.tools.securityIOTest > test_securityIO_decodeBase64Bytes FAILED
    java.lang.NoClassDefFoundError at securityIOTest.groovy:30
Running test: test test_securityIO_encodeBase64String(jervis.tools.securityIOTest)

jervis.tools.securityIOTest > test_securityIO_encodeBase64String FAILED
    java.lang.NoClassDefFoundError at securityIOTest.groovy:30
Running test: test test_securityIO_init_path_key(jervis.tools.securityIOTest)

jervis.tools.securityIOTest > test_securityIO_init_path_key FAILED
    java.lang.NoClassDefFoundError at securityIOTest.groovy:30
Running test: test test_securityIO_encodeBase64Bytes(jervis.tools.securityIOTest)

jervis.tools.securityIOTest > test_securityIO_encodeBase64Bytes FAILED
    java.lang.NoClassDefFoundError at securityIOTest.groovy:30
Running test: test test_securityIO_init_path(jervis.tools.securityIOTest)

jervis.tools.securityIOTest > test_securityIO_init_path FAILED
    java.lang.NoClassDefFoundError at securityIOTest.groovy:30
Running test: test test_securityIO_rsaEncrypt_rsaDecrypt(jervis.tools.securityIOTest)
```
Those errors in my codebase are when executing `./gradlew cobertura`.  If I execute `./gradlew check` which is how gradle runs unit tests for Groovy then all of the line numbers turn to `-1` instead of `30`.

It seems to not be importing my class when executing the unit test (or maybe it flushed my imported class).  Being explicit to import the classes in my unit tests does not do anything either.

As far as I'm concerned there's something broken about using Cobertura 2.1.1 + gradle + groovy.

If you would like to test it out then uncomment the following line in my project and run `./gradlew cobertura`.  Compare it when running it with 2.1.1 vs 2.0.3.

https://github.com/samrocketman/jervis/blob/bd36caa761daefedf1b8497215d462b75b514bb2/build.gradle#L64

I've provided as much info as I could.

**Comment by dennisl** [comment]
@samrocketman Has the Gradle plugin been updated to use Cobertura 2.1.1?

**Comment by samrocketman** [comment]
[Apparently it has.](https://github.com/stevesaliman/gradle-cobertura-plugin/blob/Release_2.2.7/CHANGELOG.md)  I'll update my dependency to use it and then report back my results.

**Comment by samrocketman** [comment]
Coverage has not improved by much for the latest versions of the Gradle plugin.

https://coveralls.io/builds/2667765

**Comment by mecorusfc** [comment]
I agree,

It looks like Cobertura 2.1.1 (saliman 2.2.7) doesn’t fix anything at all regarding coverage with Java 7 or Java 8.  Still waiting on a fix so we can move to Java 7 fully.  Right now, in my Groovy projects, I set up a separate Gradle source set that compiles with Java 6 so that I can use Cobertura to get accurate code coverage.  Once Groovy advances enough that we can’t go back to Java 6, we are stuck.

We need a solution for true coverage for Groovy/Java7/Java8.  I’m hoping that eventually Cobertura gets there.  It’s not there now.

Mike

**Comment by mecorusfc** [comment]
I should have been more specific in my response below.  For a set of projects that have 99% or 100% actual coverage, going to Java 7 under the covers dropped it to 49% like the title of the issue says.

Moving to Cobertura 2.1.1, improved that from 49% to around 80% (83% in a few projects).  So, to be fair, there is some improvement.  However, the coverage reported is still incorrect and the indications in the actual classes (lines indicated as not covered) are off as well.

Hopefully, another round of fixes to Cobertura will get us there.

Mike

From: Corum, Michael
Sent: Thursday, May 28, 2015 10:56 AM
To: 'cobertura/cobertura'; cobertura/cobertura
Subject: RE: [cobertura] Cobertura 2.0.3 - upgrading Groovy from 2.0.4 to 2.3.6 drops coverage from 98% to 49% (#184)

I agree,

It looks like Cobertura 2.1.1 (saliman 2.2.7) doesn’t fix anything at all regarding coverage with Java 7 or Java 8.  Still waiting on a fix so we can move to Java 7 fully.  Right now, in my Groovy projects, I set up a separate Gradle source set that compiles with Java 6 so that I can use Cobertura to get accurate code coverage.  Once Groovy advances enough that we can’t go back to Java 6, we are stuck.

We need a solution for true coverage for Groovy/Java7/Java8.  I’m hoping that eventually Cobertura gets there.  It’s not there now.

Mike

**Comment by samrocketman** [comment]
What's a status update for this issue?  Is there any planned work?  Currently, [my workaround][build.gradle] for code coverage reports is:

```bash
GROOVY_VERSION=1.8.9 ./gradlew clean cobertura
```

So I do have a workaround.  e.g. in my build.gradle.

```groovy
apply plugin: 'groovy'
apply plugin: 'cobertura'

ext.groovyVersion = System.getenv('GROOVY_VERSION')?: '2.4.7'

dependencies {
    compile "org.codehaus.groovy:groovy-all:${groovyVersion}"
}
...
```

[build.gradle]: https://github.com/samrocketman/jervis/blob/master/build.gradle

**Comment by christ66** [comment]
I will look into this tonight. Do you have an example scenario where the
coverage is decreased?

**Comment by samrocketman** [comment]
My code remains a constant example.

* Project: [jervis](https://github.com/samrocketman/jervis)
* [Latest Travis CI builds](https://travis-ci.org/samrocketman/jervis)

Every commit I build and do code coverage against the following versions of groovy.

```yaml
  - GROOVY_VERSION="1.8.9"
  - GROOVY_VERSION="2.0.8"
  - GROOVY_VERSION="2.1.9"
  - GROOVY_VERSION="2.2.2"
  - GROOVY_VERSION="2.3.11"
  - GROOVY_VERSION="2.4.7"
```

* [Watch the code coverage drop](https://coveralls.io/builds/6849465) for versions of Groovy above 2.0.8.  It goes from 100% coverage to low 60% for 2.1.9 and above.

**Comment by sgates** [comment]
Cobertura is missing things like groovy implicit return statements for me, but the IntelliJ profiler doesn't have an issue:

![cobertura](https://cloud.githubusercontent.com/assets/9256243/16658452/714b0182-4434-11e6-9bc8-bdded48243e0.png)
![intellij](https://cloud.githubusercontent.com/assets/9256243/16658453/714e182c-4434-11e6-9e13-0b9cb1d16558.png)

Seems related to this thread? This is the main thing causing reduced coverage % for me at the moment.

**Comment by mecorusfc** [comment]
My workaround is to have a separate Gradle source set that compiles in Java 6.  However, once I get up to a level of Groovy that won’t support that, I pretty much have to abandon Java and Groovy.

Michael Corum
VP, Technical Architecture Solutions

RGA Reinsurance Company
16600 Swingley Ridge Road
Chesterfield, Missouri 6301701706
T 636.736.7066
www.rgare.com

**Comment by mecorusfc** [comment]
Just to clear it up, all the various code coverage tools stopped working when Java 7 brought in that change that made it impossible for ASM to reverse certain byte code constructs.  Java 8 made it worse by making that transformation a required thing that you couldn’t opt out of with a –D parm.  This pretty much kills Java and Groovy for any project that does test driven development.  Oracle doesn’t seem to be listening.  Some companies are already moving away for that reason.  Newer versions of ASM have improved it slightly but I have several small projects with verified 100% code coverage that report less than 70% with JaCoCo, Clover, Cobertura, Emma and several others.

Michael Corum
VP, Technical Architecture Solutions

RGA Reinsurance Company
16600 Swingley Ridge Road
Chesterfield, Missouri 6301701706
T 636.736.7066
www.rgare.com

**Comment by blackdrag** [comment]
@mecorusfc what change is that? do you have a link so that I can read more about details?

**Comment by mecorusfc** [comment]
It’s probably going to take me several days to find it but there was a really detailed article somewhere on the web that explained the exact issue that was introduced in Java 7.  It was as a result of a performance enhancement that gave a 0.7% performance boost in cases that are almost never used every by anybody and the writer was complaining that all test coverage for all tools was lost due to this slight performance boost that nobody would see.  Apparently, it was raised as an issue with Java but Oracle said they would not fix it.

Give me a few days.  I’ll find the article.

Michael Corum
VP, Technical Architecture Solutions

RGA Reinsurance Company
16600 Swingley Ridge Road
Chesterfield, Missouri 6301701706
T 636.736.7066
www.rgare.com

**Comment by mecorusfc** [comment]
By the way, no code coverage tool has actually worked for Java 7 or Groovy over Java 7 since mid-2014 which is why some have to just make up their mind to leave Java and Groovy.

I’ll find the article.

Michael Corum
VP, Technical Architecture Solutions

RGA Reinsurance Company
16600 Swingley Ridge Road
Chesterfield, Missouri 6301701706
T 636.736.7066
www.rgare.com


From: Corum Michael <mcorum@rgare.com<mailto:mcorum@rgare.com>>
Date: Thursday, July 7, 2016 at 5:47 PM
To: cobertura/cobertura <reply@reply.github.com<mailto:reply@reply.github.com>>, cobertura/cobertura <cobertura@noreply.github.com<mailto:cobertura@noreply.github.com>>
Cc: Mention <mention@noreply.github.com<mailto:mention@noreply.github.com>>
Subject: Re: [cobertura/cobertura] Cobertura 2.0.3 - upgrading Groovy from 2.0.4 to 2.3.6 drops coverage from 98% to 49% (#184)

It’s probably going to take me several days to find it but there was a really detailed article somewhere on the web that explained the exact issue that was introduced in Java 7.  It was as a result of a performance enhancement that gave a 0.7% performance boost in cases that are almost never used every by anybody and the writer was complaining that all test coverage for all tools was lost due to this slight performance boost that nobody would see.  Apparently, it was raised as an issue with Java but Oracle said they would not fix it.

Give me a few days.  I’ll find the article.

Michael Corum
VP, Technical Architecture Solutions

RGA Reinsurance Company
16600 Swingley Ridge Road
Chesterfield, Missouri 6301701706
T 636.736.7066
www.rgare.com

**Comment by mecorusfc** [comment]
It turns out I had a bookmark for the article.  If you back through all the Cobertura issues in Jira, you’ll find other stuff on this as well.

http://chrononsystems.com/blog/java-7-design-flaw-leads-to-huge-backward-step-for-the-jvm

Note that this was announced as fixed at least once in late 2014 and twice in 2015 but the fixes never made much difference for me.  Again, the issue is Java 7 and 8 and how ASM can’t possibly keep up or even have a deterministic algorithm for figuring out byte code to code mappings now.  Later releases of ASM are better but it seems the code coverage tools are very slow to pick up the latest releases of ASM.  Even still, the best result I’ve gotten is with a patched version of JaCoCo and doing some Gradle magic to wire in a later version of ASM.  In that case, my project which has 100% coverage shows as having 76% coverage.  It’s such a small amount of code that it is easy to hand verify it.  Right now, I still have that extra sourceSet in Gradle that compiles in Java 6 in order to show correct coverage.  That won’t last for long.

Michael Corum
VP, Technical Architecture Solutions

RGA Reinsurance Company
16600 Swingley Ridge Road
Chesterfield, Missouri 6301701706
T 636.736.7066
www.rgare.com

**Comment by samrocketman** [comment]
@mecorusfc Interesting comments and article.  Gives me a sad face.

**Comment by christ66** [comment]
Yes, this is a known issue for a long time. The workaround for this though is to have the classpath used to build the class files in the `--auxClasspath` argument which cobertura uses to rebuild the one `.class` files. Makes instrumentation take a lot more time.

I was always under the impression that this was a security mechanism and not a performance issue. This helped with security because it does not allow any arbitrary user to inject byte code into a `.class` file unless they have all of the other classes which are required to compile it. I have voiced my opinion in the past about this being an issue (other oss projects contacted me about how I fixed this), and mostly got the cold shoulder.

Going back to this issue it could also be how the way groovy or java changed the way the return statements work at the byte code level. I will need to investigate what has changed.

**Comment by blackdrag** [comment]
@mecorusfc I did think about the stack frames, and remembered they have been added in Java6, but totally forget them being mandatory in Java7.  I think I do even know this blog entry you linked. I still don´t fully understand the problem cobertura has with this. Yes, transforming the bytecode for the instrumentation step is now more difficult, but the bytecode cobertura produces verifies. so this difficulty has been overcome to a satisfying extend. The branch recognition seems to be broken. So while the bytecode added is probably correct for the Verifier it does not do the right thing for a good cobertura report. And here I wonder how this is related to stack map frames.

And if stack map frames are the problem, then the problem can be reduced by using 1.5 as target bytecode level.... java and groovy allow you to do that. So all the users that have the report problem could bypass this, if it is ok for them to use such an old JDK version for their bytecode

**Comment by mecorusfc** [comment]
Yes, actually where you say 5, it is actually 6 and where you say 6, it is 7.  So Java 7 introduced the problems.  I solve it by compiling to Java 6 byte code but eventually some version of Groovy won’t work with that.  For instance, Groovy v2.3 and up currently require Java 6 as a minimum.  Eventually, they will go to requiring Java 7 as a minimum.  That’s when it gets real.

Michael Corum
VP, Technical Architecture Solutions

RGA Reinsurance Company
16600 Swingley Ridge Road
Chesterfield, Missouri 6301701706
T 636.736.7066
www.rgare.com

**Comment by blackdrag** [comment]
I said 5,because 5 does not have the problem and in 6 it is optional and not by default. Anyway.... so far the Groovy compiler can produce bytecode for JDK4-JDK8, stack map frames are required in the JDK7 and JDK8 configuration. And even if Groovy needs JDK6 atm to build and run, it can still produce bytecode for older jdk versions. There are plans for going higher with the version for the build and runtime, but not for the output bytecode version so far. So it is perfectly possible to produce JDK4 level bytecode with Groovy 2.3, 2.4 or the upcoming 2.5 (even if 2.5 would require JDK7 then)

**Comment by samrocketman** [comment]
@christ66 it sounds like `--auxClasspath` is a possible workaround based my understanding of your statement.  How would I go about implementing that in a gradle/groovy project?  Bear in mind I mostly hobby program and don't consider myself an expert in Java/Groovy.  Would you mind elaborating for me?

**Comment by sgleske-ias** [comment]
Status update: this is still an ongoing issue.  Cobertura can't handle generating valid code coverage reports against groovy beyond 2.0.8.
